## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [671]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()
        

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [672]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [673]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [674]:
# Use this space to do any additional research
# and familiarize yourself with the data.
#      Use the count function to  see if there are any missing values in the data and see column names

#      Use the dtypes function to  see type of data for each column
print(df.dtypes)



first           object
last            object
job             object
phone           object
email           object
client_id        int64
order_id         int64
order_date      object
order_week       int64
order_year       int64
item_id         object
category        object
subcategory     object
unit_price     float64
unit_cost      float64
unit_weight    float64
qty              int64
line_number      int64
dtype: object


In [676]:
# What three item categories had the most entries?
#   using value count function
df["category"].value_counts().head(3)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [677]:
# For the category with the most entries, which subcategory had the most entries?
#   look at consumables only, then use value count function 
df.loc[df['category'] == "consumables"]["subcategory"].value_counts().head(1)


subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [678]:
# Which five clients had the most entries in the data?
#   use value counts function on client_id column
df["client_id"].value_counts().head()

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [679]:
# Store the client ids of those top 5 clients in a list.
#   create list from client_id, use value counts 
# list(df["client_id"].value_counts().head().index)
top_five_clients = list(df["client_id"].value_counts().head().index)
print(top_five_clients) 

[33615, 66037, 46820, 38378, 24741]


In [680]:
# How many total units (the qty column) did the client with the most entries order order?
#   find all rows with client id 33615 and sum the qty


df.loc[df["client_id"] == 33615, "qty"].sum()





64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [681]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
#   subtotal = unit_price * qty
#   add new column "subtotal"

df['subtotal'] = df['unit_price'] * df['qty']
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number   subtotal  
0     1096.80     762.71         7.50  105            1  115164.00  
1       24.95      15.09         1.49   21            0     523.95  
2       13.52       7.86         1.68   39            6     527.28  
3       36.42      24.85         1.23   29            3    1056.18  
4      195.10     108.17        46.43   20            1    3902.00

In [683]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
#   if order_weight >50, calculate shipping price at $7 * unit_weight
#   if order_weight <=50, $10 * unit weight
#   add new column "shipping_price"

# df['shipping_price'] = df.apply(lambda df: 7 * (row['unit_weight'] * row['qty']) if (row['unit_weight'] * row['qty']) > 50 else 10 * (row['unit_weight'] * row['qty']))
# df.head()
df['order_weight'] = df['unit_weight'] * df['qty']
df['shipping_price'] = df.apply(lambda row: 7 * row['order_weight'] if row['order_weight'] > 50 else 10 * row['order_weight'], axis=1)
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...     category         subcategory unit_price  \
0          17        2023  ...        decor            wall art    1096.80   
1          20        2023  ...  consumables                pens      24.95   
2           4        2023  ...     software  project management      13.52   
3          17        2023  ...  consumables                pens      36.42   
4          21        2023  ...  consumables                misc     195.10   

   unit_cost  unit_weight  qty  line_number   subtotal  order_weight  \
0     762.71         7.50  105            1  115164.00        787.50   
1      15.09         1.49   21            0     523.95         31.29   
2       7.86         1.68   39            6     527.28         65.52   
3      24.85         1.23   29            3    1056.18         35.67   
4     108.17        46.43   20            1    3902.00        928.60   

   shipping_price  
0         5512.50  
1          312.90  
2          458.64  
3          356.70  
4         6500.20  

[5 rows x 21 columns]

In [684]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%


df['total_price'] = df['total_price'] = (df['subtotal'] + df['shipping_price']) * 1.0925
df['total_price'] = df['total_price'].round(2)
df.head()



first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...         subcategory unit_price unit_cost  \
0          17        2023  ...            wall art    1096.80    762.71   
1          20        2023  ...                pens      24.95     15.09   
2           4        2023  ...  project management      13.52      7.86   
3          17        2023  ...                pens      36.42     24.85   
4          21        2023  ...                misc     195.10    108.17   

   unit_weight  qty  line_number   subtotal  order_weight  shipping_price  \
0         7.50  105            1  115164.00        787.50         5512.50   
1         1.49   21            0     523.95         31.29          312.90   
2         1.68   39            6     527.28         65.52          458.64   
3         1.23   29            3    1056.18         35.67          356.70   
4        46.43   20            1    3902.00        928.60         6500.20   

   total_price  
0    131839.08  
1       914.26  
2      1077.12  
3      1543.57  
4     11364.40  

[5 rows x 22 columns]

In [685]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
#   create new column "line_cost"
#   line_cost=(unit_cost+shipping_price)*qty
df['line_cost'] = (df['unit_cost'] * df['qty']) + df['shipping_price']
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number   subtotal  order_weight  shipping_price  total_price  \
0            1  115164.00        787.50         5512.50    131839.08   
1            0     523.95         31.29          312.90       914.26   
2            6     527.28         65.52          458.64      1077.12   
3            3    1056.18         35.67          356.70      1543.57   
4            1    3902.00        928.60         6500.20     11364.40   

   line_cost  
0   85597.05  
1     629.79  
2     765.18  
3    1077.35  
4    8663.60  

[5 rows x 23 columns]

In [686]:
# Create a column for the profit of each line using line cost and line price
#   create new column "profit"
#   profit = subtotal- line_cost
df['profit'] = df['total_price'] - df['line_cost']
df = df.round(2)
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   
3          17        2023  ...     24.85        1.23   29            3   
4          21        2023  ...    108.17       46.43   20            1   

    subtotal  order_weight  shipping_price  total_price  line_cost    profit  
0  115164.00        787.50         5512.50    131839.08   85597.05  46242.03  
1     523.95         31.29          312.90       914.26     629.79    284.47  
2     527.28         65.52          458.64      1077.12     765.18    311.94  
3    1056.18         35.67          356.70      1543.57    1077.35    466.22  
4    3902.00        928.60         6500.20     11364.40    8663.60   2700.80  

[5 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [695]:
# Check your work using the totals above

order_2742071 = df[df['order_id'] == 2742071]
total_price_2742071 = order_2742071['total_price'].sum()
print(f"order 2742071 total: ${total_price_2742071:.2f}")

order_2173913 = df[df['order_id'] == 2173913]
total_price_2173913 = order_2173913['total_price'].sum()
print(f"order 2173913 total: ${total_price_2173913:.2f}")

order_6128929 = df[df['order_id'] == 6128929]
total_price_6128929 = order_6128929['total_price'].sum()
print(f"order 6128929 total: ${total_price_6128929:.2f}")



order 2742071 total: $152811.89
order 2173913 total: $162388.71
order 6128929 total: $923441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [694]:

top_clients_data = df[df['client_id'].isin(top_five_clients)]

# Calculate the sum of total_price for each top client
total_price_sum_per_client = top_clients_data.groupby('client_id')['total_price'].sum().reset_index()

# Convert total_price to numeric type
total_price_sum_per_client['total_price'] = total_price_sum_per_client['total_price'].astype(float)

# Format the currency values for better display
total_price_sum_per_client['total_price'] = total_price_sum_per_client['total_price'].map('${:,.2f}'.format)

# Displaying the sum of total_price for each top client
print(total_price_sum_per_client)

   client_id     total_price
0      24741  $82,268,892.02
1      33615   $8,377,308.52
2      38378  $12,906,550.87
3      46820   $9,743,794.36
4      66037  $10,259,514.79


In [706]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 

summary_df = df[df['client_id'].isin(top_five_clients)]
summed_values = summary_df.groupby('client_id').agg({
    'qty': 'sum',
    'shipping_price': 'sum',
    'total_price': 'sum',
    'line_cost': 'sum',
    'profit': 'sum'
}).reset_index()

# filtered_orders
summed_values.head()



client_id     qty  shipping_price  total_price    line_cost       profit
0      24741  239862      5126448.37  82268892.02  45688899.71  36579992.31
1      33615   64313      1828984.89   8377308.52   6175313.91   2201994.61
2      38378   73667      3429455.40  12906550.87   9634720.98   3271829.89
3      46820   75768      1601448.84   9743794.36   7007191.64   2736602.72
4      66037   43018      1395151.85  10259514.79   7004482.98   3255031.81

In [731]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 


# Define a function that converts a dollar amount to millions.


# Apply the currency_format_millions function to only the money columns. 


# Rename the columns to reflect the change in the money format. 



summed_values_formatted = summed_values.rename(columns={'client_id':"Client ID",
                                                        'qty':'Units',
                                                        'shipping_price':'Shipping (millions)',
                                                        'total_price':'Total Revenue (millions)',
                                                        'line_cost':'Total Cost (millions)',
                                                        'profit':'Total Profit (millions)'})
summed_values_formatted.head()

def convert_to_millions(x):
    return '${:,.2f}M'.format(x / 1000000)

money_columns = ['Shipping (millions)', 'Total Revenue (millions)', 'Total Cost (millions)', 'Total Profit (millions)']

summed_values_formatted[money_columns] = summed_values_formatted[money_columns].apply(lambda x: x.apply(convert_to_millions))

summed_values_formatted




Client ID   Units Shipping (millions) Total Revenue (millions)  \
0      24741  239862     $5.13M             $82.27M                 
1      33615   64313     $1.83M              $8.38M                 
2      38378   73667     $3.43M             $12.91M                 
3      46820   75768     $1.60M              $9.74M                 
4      66037   43018     $1.40M             $10.26M                 

  Total Cost (millions) Total Profit (millions)  
0    $45.69M               $36.58M               
1     $6.18M                $2.20M               
2     $9.63M                $3.27M               
3     $7.01M                $2.74M               
4     $7.00M                $3.26M

In [732]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
formatted_sorted = summed_values_formatted.sort_values('Total Profit (millions)', ascending=False)
formatted_sorted

Client ID   Units Shipping (millions) Total Revenue (millions)  \
0      24741  239862     $5.13M             $82.27M                 
2      38378   73667     $3.43M             $12.91M                 
4      66037   43018     $1.40M             $10.26M                 
3      46820   75768     $1.60M              $9.74M                 
1      33615   64313     $1.83M              $8.38M                 

  Total Cost (millions) Total Profit (millions)  
0    $45.69M               $36.58M               
2     $9.63M                $3.27M               
4     $7.00M                $3.26M               
3     $7.01M                $2.74M               
1     $6.18M                $2.20M

####    The top five most profitable clients generated combined revenues of 123.56 million (USD) and a net profit of 48.05 million(USD). There are two key observations revealed in this analysis: 
#####   The averages for Total Revenue, Total Cost, and Total Profit are heavily weighted by the most profitable client, 24741;this needs to be considered in any subsequent analyses of the full data to ensure that reporting averages is not skewed  by the results of a single client.
#####   The Cost to Revenue ratio for client 24741 is significantly lower in comparison to the remaining top four clients; further analysis is reccomended to understand why they are so much more profitable in relation to the other top clients.
      


